crates/wasm_interp/src/execute.rs

```diff
@@ -681,36 +738,96 @@ impl<'a> ExecutionState<'a> {
             }
             I32MUL => {
-                let x = self.value_stack.pop_i32();
-                let y = self.value_stack.pop_i32();
-                self.value_stack.push(Value::I32(y * x));
-            }
-            I32DIVS => todo!("{:?} @ {:#x}", op_code, file_offset),
-            I32DIVU => todo!("{:?} @ {:#x}", op_code, file_offset),
-            I32REMS => todo!("{:?} @ {:#x}", op_code, file_offset),
-            I32REMU => todo!("{:?} @ {:#x}", op_code, file_offset),
-            I32AND => todo!("{:?} @ {:#x}", op_code, file_offset),
-            I32OR => todo!("{:?} @ {:#x}", op_code, file_offset),
-            I32XOR => todo!("{:?} @ {:#x}", op_code, file_offset),
-            I32SHL => todo!("{:?} @ {:#x}", op_code, file_offset),
-            I32SHRS => todo!("{:?} @ {:#x}", op_code, file_offset),
-            I32SHRU => todo!("{:?} @ {:#x}", op_code, file_offset),
-            I32ROTL => todo!("{:?} @ {:#x}", op_code, file_offset),
-            I32ROTR => todo!("{:?} @ {:#x}", op_code, file_offset),
+                let arg2 = self.value_stack.pop_i32();
+                let arg1 = self.value_stack.pop_i32();
+                self.value_stack.push(Value::from(arg1.wrapping_mul(arg2)));
+            }
+            I32DIVS => {
+                let arg2 = self.value_stack.pop_i32();
+                let arg1 = self.value_stack.pop_i32();
+                self.value_stack.push(Value::from(arg1.wrapping_div(arg2)));
+            }
+            I32DIVU => {
+                let arg2 = self.value_stack.pop_u32();
+                let arg1 = self.value_stack.pop_u32();
+                self.value_stack.push(Value::from(arg1.wrapping_div(arg2)));
+            }
+            I32REMS => {
+                let arg2 = self.value_stack.pop_i32();
+                let arg1 = self.value_stack.pop_i32();
+                self.value_stack.push(Value::from(arg1.wrapping_rem(arg2)));
+            }
+            I32REMU => {
+                let arg2 = self.value_stack.pop_u32();
+                let arg1 = self.value_stack.pop_u32();
+                self.value_stack.push(Value::from(arg1.wrapping_rem(arg2)));
+            }
+            I32AND => {
+                let arg2 = self.value_stack.pop_u32();
+                let arg1 = self.value_stack.pop_u32();
+                self.value_stack.push(Value::from(arg1 & arg2));
+            }
+            I32OR => {
+                let arg2 = self.value_stack.pop_u32();
+                let arg1 = self.value_stack.pop_u32();
+                self.value_stack.push(Value::from(arg1 | arg2));
+            }
+            I32XOR => {
+                let arg2 = self.value_stack.pop_u32();
+                let arg1 = self.value_stack.pop_u32();
+                self.value_stack.push(Value::from(arg1 ^ arg2));
+            }
+            I32SHL => {
+                let arg2 = self.value_stack.pop_u32();
+                let arg1 = self.value_stack.pop_u32();
+                let k = arg2 % 32;
```

Comment:
maybe  `k = u32::min(arg2, 32)` is a bit clearer? or add a comment about how wasm handles bitshifts by more bits than the type has.

Reply:
Ok the spec explicitly says remainder so min would be the wrong behaviour. I'll add a comment.